File: py-algorithm/arrays/check_if_binary_string_has_at_most_one_segment_of_ones.py
import unittest

from math import gcd
from fractions import Fraction
from typing import List
from queue import PriorityQueue
from collections import Counter
import itertools

"""
    https://leetcode.com/problems/check-if-binary-string-has-at-most-one-segment-of-ones/
    
    Cho một chuỗi nh phân không có số 0 ở đầu. Trả về TRUE nếu có nhiều nhất một đoạn liên
    tiếp các số 1, ngược lại trả về FALSE.

    Ví dụ:

        s	        Kết quả	    Giải thích
        111000	    true	    chỉ có 1 đoạn 111
        110011	    false	    có 2 đoạn 11
        1	        true	    một đoạn 1
        101	        false	    hai đoạn 1
"""


class Solution(unittest.TestCase):

    def checkOnesSegment_OK(self, s: str) -> bool:
        return not s.__contains__('01')

    def checkOnesSegment(self, s: str) -> bool:
        count = 0
        for i in range(len(s) - 1):
            if s[i] != s[i + 1]:
                count += 1
                if count > 1:
                    return False

        return True

    def test_1(self):
        test_cases = [
            ("110011", False),
            ("110", True),
            ("1001", False),
            ("1", True),
            ("10", True),
            ("111", True),
            ("1000", True),
        ]
        for s, expected in test_cases:
            self.assertEqual(expected, self.checkOnesSegment(s))


if __name__ == '__main__':
    unittest.main()

File: go_algorithm/arrays/check_if_binary_string_has_at_most_one_segment_of_ones.go
package main

import (
	"go_algorithm/test_util"
	"strings"
)

/*
	https://leetcode.com/problems/check-if-binary-string-has-at-most-one-segment-of-ones/

	Cho một chuỗi nhị phân s không có số 0 ở đầu.

	Hãy trả về true nếu s chứa nhiều nhất một đoạn liên tiếp các số 1. Ngược lại, trả về false.
 */
func checkOnesSegment_OK(s string) bool {
	return !strings.Contains(s, "01")
}

func checkOnesSegment(s string) bool {
	count := 0
	for i := 0; i < len(s) - 1; i++ {
		if s[i] != s[i + 1] {
			count++
			if count > 1 {
				return false
			}
		}
	}
	return true
}

func main() {
	testCases := map[string]bool{ //
		"110011": false,
		"110":    true,
		"1001":   false,
		"1":      true,
		"10":     true,
		"111":    true,
		"1000":   false,
	}
	for input, expected := range testCases {
		test_util.AssertBool(expected, checkOnesSegment(input))
	}

}
